packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -519,6 +520,18 @@ abstract class ImageStreamCompleter with Diagnosticable {
     _checkDisposed();
     _hadAtLeastOneListener = true;
     _listeners.add(listener);
+    _synchronouslyCallAddedListener(listener);
+  }
+
+  /// Add a listener for errors. It is similar to [addListener], but see
```

Comment:
> Add a listener for errors

From the looks of the code this listener doesn't just fire for errors. It also fires for onImage and onChunk events. Is that intentional? Should it only fire on errors? Again, this behavior needs to be properly documented!

Reply:
FWIW, I think it would be fine if this API was modified so that it _only_ fired on errors, and if it dropped all these listeners once the stream succeeded in producing an image.

We'd just hav eto edocument it that way.

Also, sorry @fzyzcjy - I didn't do a very close pass on documentation when I was reviewing :(

Reply:
> FWIW, I think it would be fine if this API was modified so that it only fired on errors, and if it dropped all these listeners once the stream succeeded in producing an image.

That looks like another possible approach! So now I have two different ways in my mind:

1. (What is currently implemented) Let ResizeImage remove the (peek) listener when `onImage`, see https://github.com/fzyzcjy/flutter/blob/ccac069c5bc4392e6c23281f2432b50b3f7e6d59/packages/flutter/lib/src/painting/image_provider.dart#L1450
2. (What @dnfield suggests here) Let ImageStreamCompleter does the job.

Approach 1 has pros that it is more flexible for users of ImageStreamCompleter, and peek listener is symmetric to the normal listener.

Approach 2 has pros that ImageStreamCompleter expose a smaller API.

I am happy to implement either thing. I will just wait for your opinions before continuing!

Reply:
So... Which one should I implement? :eyes:

Reply:
I'd favor the one that exposes less new API surface. A peek listener might be easier to misuse. Something that listens either for an error or until an image is completed seems easier to think about when to use.

Reply:
I see. I will implement that!